### ISSUE TITLE
memory leak: chunked data array re-added to DataConnection._chunkedData map after receiving all chunks

### ISSUE THREAD

**Issue by sindelaru** [peers/peerjs#369]
I've encountered an issue in DataConnection.prototype._handleDataMessage, which becomes obvious when receiving many chunked messages.

In the current code, after all chunks are received, the _chunkid_ map key is deleted, a Blob is created with all the concatenated chunk data and the function is recursively called again, in order to fire a "data" event to outside listeners.

However after returning from the recursive call, the key is once again added to the map, which causes data chunks to accumulate and never be released. I am not sure why that would be the actual intention. Moving the "add to map" instruction **before** the "all chunks here"-check seems to fix the memory issue.

Any comments on whether the fix is correct are welcome - otherwise - I hope it helps...

```
if (data.__peerData) {

    var id = data.__peerData;
    var chunkInfo = this._chunkedData[id] || {data: [], count: 0, total: data.total};

    chunkInfo.data[data.n] = data.data;
    chunkInfo.count += 1;
    this._chunkedData[id] = chunkInfo;

    if (chunkInfo.total === chunkInfo.count) {
      // Clean up before making the recursive call to `_handleDataMessage`.
      delete this._chunkedData[id];

      // We've received all the chunks--time to construct the complete data.
      data = new Blob(chunkInfo.data);
      this._handleDataMessage({data: data});
    }

    return;
  }

```

**Comment by balintx99** [comment]
@sindelaru Man, you totally saved my day :D
I was experiencing a huge memory leak when sending large files in (10mb chunks). And swaping these two lines, just eliminated the issue completely.

@afrokick I don't know whether this solution still under investigation is or not, but please implement it in the next version. It is a total life saver.

**Comment by afrokick** [comment]
Fixed in #556 We'll release it in 1.0.2
